add clipping to square-distance calculation (#165)
rsokl/Learning_Python: Python/Module3_IntroducingNumpy/Broadcasting.md

```python
# computing all the distances
>>> dists = np.sqrt(x_y_sqrd - x_y_prod)
>>> dists.shape
(5, 6)
```

> #### A Subtle Issue with Floating Point Precision

Reviewer: nit:
```suggestion
#### A Subtle Issue with Floating-point Precision
```
or
```suggestion
#### A Subtle Issue with Floating-Point Precision
```

There is one more important and very subtle detail that we have to deal with.
In terms of pure mathematics, `x_y_sqrd - x_y_prod` must be a strictly non-negative value (i.e. its smallest possible value is $0$), since it is equivalent to

\begin{equation}
\sum_{i=0}^{D-1}{(x_{i} - y_{i})^2}
\end{equation}

> That being said, are working with floating point numbers, which do not always behave exactly like rational numbers when we do arithmetic on them.

Reviewer: nit:
```suggestion
That being said, are working with floating-point numbers, which do not always behave exactly like rational numbers when we do arithmetic with them.
```

> Indeed, we saw earlier that the [quirks of floating point arithmetic](https://www.pythonlikeyoumeanit.com/Module2_EssentialsOfPython/Basic_Objects.html#Understanding-Numerical-Precision) can lead to surprising results. 

Reviewer: nit:
```suggestion
Indeed, we saw earlier that the [quirks of floating-point arithmetic](https://www.pythonlikeyoumeanit.com/Module2_EssentialsOfPython/Basic_Objects.html#Understanding-Numerical-Precision) can lead to surprising results. 
```

Here, the strange behavior is that `x_y_sqrd - x_y_prod` **can produce negative numbers**!

Let's see this in action.
We'll take the following shape-(2, 3) array

```python
# A carefully-selected input that will trigger surprising

> # floating point precision issues in our distance calculation

Reviewer: nit:
```suggestion
# floating-point precision issues in our distance calculation
```

>>> x = np.array([[4.700867387959219, 4.700867387959219, 4.700867387959219],
...               [4.700867387959219, 4.700867387959219, 4.700867387959219]])
```

> And compute the square distance between all combination of rows `x`; the result should simply a shape-(2, 2) array of zeros.

Reviewer: ```suggestion
And compute the square distance between all combinations of rows `x`; the result should simply be a shape-(2, 2) array of zeros.
```

```python
# The square-distances should be exactly zero, but they
# will instead be (*very* small) negative numbers!
>>> sqr_dists = -2 * np.matmul(x, x.T)  # `x_x_prod`
>>> sqr_dists +=  np.sum(x**2, axis=1)[:, np.newaxis]
>>> sqr_dists += np.sum(x**2, axis=1)
>>> sqr_dists
array([[-2.842170943040401e-14, -2.842170943040401e-14],
       [-2.842170943040401e-14, -2.842170943040401e-14]])
```

> These values are *very* close to being zero, so it is not as if the magnitude of our result is wildly off, but the critical issue here is that the quirks of floating point arithmetic produced (very small) negative numbers.

Reviewer: nit:
```suggestion
These values are *very* close to being zero, so it is not as if the magnitude of our result is wildly off, but the critical issue here is that the quirks of floating-point arithmetic produced (very small) negative numbers.
```

```python
# Taking the square-root of negative floats will produce NaNs
>>> np.sqrt(sqr_dists)
array([[nan, nan],
       [nan, nan]])
```

`nan` stands for "Not a Number", which is to say that the square-root of a negative number cannot produce a real-valued float (the result will be an [imaginary number](https://www.pythonlikeyoumeanit.com/Module2_EssentialsOfPython/Basic_Objects.html#Complex-Numbers), but NumPy will not swap out real numbers for imaginary ones unless we explicitly permit it to).

The solution to this problem comes from the realization that this occurrence is truly an edge-case, and that it will only manifest when the result *ought* to have been zero and indeed is very close to zero.
Thus we can [clip](https://numpy.org/doc/stable/reference/generated/numpy.clip.html) `sqr_dists` such that any value that falls below zero gets set to zero and all other values will be left alone:

```python
# "Clipping" all negative entries so that they are set to 0 
>>> np.sqrt(np.clip(sqr_dists, a_min=0, a_max=None))
array([[0., 0.],
       [0., 0.]])
```
<!-- #endregion -->

<!-- #region -->
#### The Final Answer, At Last!

> In total, we have successfully used vectorization to compute the all pairs of distances, while only requiring an array of shape-$(M, N)$ to do so! This is the memory-efficient, vectorized form – the stuff that dreams are made of.

Reviewer: ```suggestion
In total, we have successfully used vectorization to compute all the pairs of distances, while only requiring an array of shape-$(M, N)$ to do so! This is the memory-efficient, vectorized form – the stuff that dreams are made of.
```

Let's write the function that performs this computation in full.  

```python
def pairwise_dists(x, y):
    """ Computing pairwise distances using memory-efficient 
    vectorization.

    Parameters
    ----------
    x : numpy.ndarray, shape=(M, D)
    y : numpy.ndarray, shape=(N, D)

    Returns
    -------
    numpy.ndarray, shape=(M, N)
        The Euclidean distance between each pair of

>         rows between `x` and `y`."""

Reviewer: do you want to leave the end-quote on the same line or a newline? not sure what the consistent thing is off the top of my head

Author: I think same line is what `black` does. So I'll go with it.
